frame/staking/src/voter_bags.rs

```diff
@@ -0,0 +1,328 @@
+			if bag.head.as_ref() == Some(voter_id) {
+				bag.head = node.next;
+			}
+			if bag.tail.as_ref() == Some(voter_id) {
+				bag.tail = node.prev;
+			}
+
+			// now get rid of the node itself
+			crate::VoterNodes::<T>::remove(node.bag_idx, voter_id);
+		}
+
+		for (_, bag) in bags {
+			bag.put();
+		}
+	}
+}
+
+/// A Bag is a doubly-linked list of voters.
+///
+/// Note that we maintain both head and tail pointers. While it would be possible to get away
+/// with maintaining only a head pointer and cons-ing elements onto the front of the list, it's
+/// more desirable to ensure that there is some element of first-come, first-serve to the list's
+/// iteration so that there's no incentive to churn voter positioning to improve the chances of
+/// appearing within the voter set.
+#[derive(DefaultNoBound, Encode, Decode)]
+pub struct Bag<T: Config> {
+	head: Option<AccountIdOf<T>>,
+	tail: Option<AccountIdOf<T>>,
+
+	#[codec(skip)]
+	bag_idx: BagIdx,
+}
+
+impl<T: Config> Bag<T> {
+	/// Get a bag by idx.
+	pub fn get(bag_idx: BagIdx) -> Option<Bag<T>> {
+		crate::VoterBags::<T>::try_get(bag_idx).ok().map(|mut bag| {
+			bag.bag_idx = bag_idx;
+			bag
+		})
+	}
+
+	/// Get a bag by idx or make it, appropriately initialized.
+	pub fn get_or_make(bag_idx: BagIdx) -> Bag<T> {
+		Self::get(bag_idx).unwrap_or(Bag { bag_idx, ..Default::default() })
+	}
+
+	/// Put the bag back into storage.
+	pub fn put(self) {
+		crate::VoterBags::<T>::insert(self.bag_idx, self);
+	}
+
+	/// Get the head node in this bag.
+	pub fn head(&self) -> Option<Node<T>> {
+		self.head.as_ref().and_then(|id| Node::get(self.bag_idx, id))
+	}
+
+	/// Get the tail node in this bag.
+	pub fn tail(&self) -> Option<Node<T>> {
+		self.tail.as_ref().and_then(|id| Node::get(self.bag_idx, id))
+	}
+
+	/// Iterate over the nodes in this bag.
+	pub fn iter(&self) -> impl Iterator<Item = Node<T>> {
+		sp_std::iter::successors(self.head(), |prev| prev.next())
+	}
+
+	/// Insert a new voter into this bag.
+	///
+	/// This is private on purpose because it's naive: it doesn't check whether this is the
+	/// appropriate bag for this voter at all. Generally, use [`VoterList::insert`] instead.
+	///
+	/// Storage note: this modifies storage, but only for the nodes. You still need to call
+	/// `self.put()` after use.
+	fn insert(&mut self, voter: VoterOf<T>) {
+		let id = voter.id.clone();
+		let tail = self.tail();
+
+		// insert the actual voter
+		let voter_node = Node::<T> {
+			voter,
+			prev: tail.as_ref().map(|prev| prev.voter.id.clone()),
+			next: None,
+			bag_idx: self.bag_idx,
+		};
+		voter_node.put();
+
+		// update the previous tail
+		if let Some(mut tail) = tail {
+			tail.next = Some(id.clone());
+			tail.put();
+		}
+
+		// update the internal bag links
+		if self.head.is_none() {
+			self.head = Some(id.clone());
+		}
+		self.tail = Some(id);
+	}
+}
+
+/// A Node is the fundamental element comprising the singly-linked lists which for each bag.
```

Comment:
It looks rather like a doubly-linked list to me. We need that to make removals fast, right?

Reply:
Yeah, I started writing it as a singly-linked list, then had to double it for removals. Good spot, I'll update the comment.

Reply:
https://github.com/paritytech/substrate/pull/9081/commits/1e7dfd96a898eb2e01e430625e0cc635f1539682